p2panda-rs/src/next/operation/validate.rs

```diff
@@ -0,0 +1,338 @@
+///                     │            Look-Up   │   │    │          │          │
+///                     ▼                      │   │    │          │          │
+///              ┌──────────────┐    ┌──────┐  │   │    │          │          │
+///              │PlainOperation│    │Schema│  │   │    │          │          │
+///              └──────┬───────┘    └──┬───┘  │   │    │          │          │
+///                     │               │      │   │    │          │          │
+///                     │               │      │   │    │          │          │
+///                     │               │      │   │    │          │          │
+///                     │               │      │   │    │          │          │
+///                     │               ▼      ▼   ▼    ▼          ▼          │
+///                     └───────────►  validate_operation_and_entry() ◄───────┘
+///                                                 │
+///                                                 │
+///                                                 │
+///                                                 │
+///                                                 ▼
+///                                         ┌─────────────────┐
+///                                         │VerifiedOperation│
+///                                         └─────────────────┘
+/// ```
+#[allow(clippy::too_many_arguments)]
+pub fn validate_operation_with_entry(
+    entry: &Entry,
+    entry_encoded: &EncodedEntry,
+    skiplink_entry: Option<&Entry>,
+    skiplink_entry_hash: Option<&Hash>,
+    backlink_entry: Option<&Entry>,
+    backlink_entry_hash: Option<&Hash>,
+    plain_operation: &PlainOperation,
+    operation_encoded: &EncodedOperation,
+    schema: &Schema,
+) -> Result<VerifiedOperation, VerifiedOperationError> {
+    // Verify that the entry belongs to this operation
+    validate_payload(entry, operation_encoded)?;
+
+    // Verify that the entries links are correct
+    validate_log_integrity(
+        entry,
+        skiplink_entry,
+        skiplink_entry_hash,
+        backlink_entry,
+        backlink_entry_hash,
+    )?;
+
+    // The operation id is the result of a hashing function over the entry bytes.
+    let operation_id = entry_encoded.hash().into();
+
+    // Validate and convert plain operation with the help of a schema
+    let operation = validate_operation(plain_operation, schema)?;
+
+    Ok(VerifiedOperation {
+        entry: entry.to_owned(),
+        operation,
+        operation_id,
+    })
+}
+
+/// Checks the fields of an operation-like data type against a schema.
+///
+/// This method checks against:
+///
+/// 1. Correct operation format (#OP2)
+/// 2. Correct canonic operation field values, like document view ids of pinned relations (no
+///    duplicates, sorted) (#OP3)
+/// 3. Schema matches the given operation fields (#OP4)
+pub fn validate_operation<O: Actionable + Schematic>(
```

Comment:
Is this the only place these traits are used?

Reply:
Yes!